common/src/main/java/com/microsoft/identity/common/internal/cache/SharedPreferencesAuthorityValidationMetadataCache.java

```diff
@@ -0,0 +1,84 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package com.microsoft.identity.common.internal.cache;
+
+import android.content.Context;
+import androidx.annotation.Nullable;
+import androidx.annotation.VisibleForTesting;
+import com.microsoft.identity.common.adal.internal.cache.IStorageHelper;
+import com.microsoft.identity.common.adal.internal.cache.StorageHelper;
+
+public class SharedPreferencesAuthorityValidationMetadataCache implements IDefaultAuthorityValidationMetadataCache {
+
+    private static final String CACHE_VALUE_SEPARATOR = "-";
+    private static final String AUTHORITY_VALIDATION_METADATA_CACHE_GUID = "33DD5583-1098-4617-AF07-2D327BC4C0E4";
+
+    /**
+     * The name of the SharedPreferences file on disk.
+     * This file is only used to store MSAL CPP authority validation metadata
+     */
+    private static final String DEFAULT_CPP_AUTHORITY_VALIDATION_METADATA_SHARED_PREFERENCES =
+            "com.microsoft.identity.client.cpp_authority_validation_metadata";
+
+    private final ICacheKeyValueDelegate mCacheKeyValueDelegate;
+
+    private final ISharedPreferencesFileManager mCppAuthorityValidationMetadataSharedPreferencesFileManager;
+
+    public SharedPreferencesAuthorityValidationMetadataCache(final Context context) {
+        mCacheKeyValueDelegate = new CacheKeyValueDelegate();
+
+        // We don't need to encrypt this cache file, since the cache file does not contain any
+        // secrets and other apps do not have access to this cache file.
+        final IStorageHelper storageHelper = null;
+        mCppAuthorityValidationMetadataSharedPreferencesFileManager =
+                SharedPreferencesFileManager.getSharedPreferences(
+                        context,
+                        DEFAULT_CPP_AUTHORITY_VALIDATION_METADATA_SHARED_PREFERENCES,
+                        storageHelper
+                );
+    }
+
+    @Override
+    public String generateAuthorityValidationMetadataKey(final String environment) {
+        return AUTHORITY_VALIDATION_METADATA_CACHE_GUID + CACHE_VALUE_SEPARATOR + environment;
+    }
+
+    @Override
+    public void saveAuthorityValidationMetadata(final String environment, final String cacheValue) {
+        final String cacheKey = generateAuthorityValidationMetadataKey(environment);
+        mCppAuthorityValidationMetadataSharedPreferencesFileManager.putString(cacheKey, cacheValue);
+    }
+
+    @Override
+    @Nullable
+    public String getAuthorityValidationMetadata(final String environment) {
+        final String cacheKey = generateAuthorityValidationMetadataKey(environment);
+        return mCppAuthorityValidationMetadataSharedPreferencesFileManager.getString(cacheKey);
+    }
+
+    @Override
+    @VisibleForTesting(otherwise = VisibleForTesting.NONE)
+    public synchronized void clearCache() {

```

Comment:
I understand that was some concern about a cache poisoning scenario? Would it ever make sense to call this method outside of the context of a test?

Reply:
This is not related to the cache poisoning scenario, this function should only be called in tests